torchx/examples/mnist.exs

```diff
@@ -0,0 +1,163 @@
+        File.read!(path)
+      else
+        IO.puts("Fetching #{zip} from https://storage.googleapis.com/cvdf-datasets/mnist/\n")
+        :inets.start()
+        :ssl.start()
+
+        {:ok, {_status, _response, data}} = :httpc.request(:get, {base_url ++ zip, []}, [], [])
+        File.mkdir_p!("tmp")
+        File.write!(path, data)
+
+        data
+      end
+
+    :zlib.gunzip(data)
+  end
+
+  def download(images, labels) do
+    <<_::32, n_images::32, n_rows::32, n_cols::32, images::binary>> =
+      unzip_cache_or_download(images)
+
+    train_images =
+      images
+      |> Nx.from_binary({:u, 8})
+      |> Nx.reshape({n_images, n_rows * n_cols}, names: [:batch, :input])
+      |> Nx.divide(255)
+      |> Nx.to_batched_list(30)
+
+    IO.puts("#{n_images} #{n_rows}x#{n_cols} images\n")
+
+    <<_::32, n_labels::32, labels::binary>> = unzip_cache_or_download(labels)
+
+    train_labels =
+      labels
+      |> Nx.from_binary({:u, 8})
+      |> Nx.reshape({n_labels, 1}, names: [:batch, :output])
+      |> Nx.equal(Nx.tensor(Enum.to_list(0..9)))
+      |> Nx.to_batched_list(30)
+
+    IO.puts("#{n_labels} labels\n")
+
+    {train_images, train_labels}
+  end
+
+  def train_epoch(cur_params, imgs, labels) do
+    total_batches = Enum.count(imgs)
+
+    imgs
+    |> Enum.zip(labels)
+    |> Enum.reduce({cur_params, Nx.tensor(0.0), Nx.tensor(0.0)}, fn
+      {imgs, tar}, {cur_params, avg_loss, avg_accuracy} ->
+        update_with_averages(cur_params, imgs, tar, avg_loss, avg_accuracy, total_batches)
+    end)
+  end
+
+  def train(imgs, labels, params, opts \\ []) do
+    epochs = opts[:epochs] || 5
+
+    for epoch <- 1..epochs, reduce: params do
+      cur_params ->
+        {time, {new_params, epoch_avg_loss, epoch_avg_acc}} =
+          :timer.tc(__MODULE__, :train_epoch, [cur_params, imgs, labels])
+
+        epoch_avg_loss =
+          epoch_avg_loss
+          |> Nx.backend_transfer()
+          |> Nx.to_scalar()
+
+        epoch_avg_acc =
+          epoch_avg_acc
+          |> Nx.backend_transfer()
+          |> Nx.to_scalar()
+
+        IO.puts("Epoch #{epoch} Time: #{time / 1_000_000}s")
+        IO.puts("Epoch #{epoch} average loss: #{inspect(epoch_avg_loss)}")
+        IO.puts("Epoch #{epoch} average accuracy: #{inspect(epoch_avg_acc)}")
+        IO.puts("\n")
+        new_params
+    end
+  end
+end
+
+Nx.default_backend(Torchx.Backend)
+
+{train_images, train_labels} =
+  MNIST.download('train-images-idx3-ubyte.gz', 'train-labels-idx1-ubyte.gz')
+
+IO.puts("Initializing parameters...\n")
+params = MNIST.init_random_params()
+
+IO.puts("Training MNIST for 10 epochs...\n\n")
+final_params = MNIST.train(train_images, train_labels, params, epochs: 10)
+
+IO.puts("The result of the first batch")
+IO.inspect(MNIST.predict(final_params, hd(train_images)) |> Nx.argmax(axis: :output))
+
+IO.puts("Labels for the first batch")
+IO.inspect(hd(train_labels) |> Nx.as_type({:s, 8}) |> Nx.argmax(axis: :output))
```

Comment:
We shouldn't need this conversion here either. :)

Reply:
Same reason as above.
Should we add type check in Torchx argmax and issue a warning or raise more imperative error message?